src/api/altorenderer.cpp

```diff
@@ -0,0 +1,252 @@
+        "\t\t\t\t<processingSoftware>\n"
+        "\t\t\t\t\t<softwareName>tesseract ");
+        AppendString(TessBaseAPI::Version());
+        AppendString("</softwareName>\n"
+        "\t\t\t\t</processingSoftware>\n"
+        "\t\t\t</ocrProcessingStep>\n"
+        "\t\t</OCRProcessing>\n"
+        "\t</Description>\n"
+        "\t<Layout>\n");
+
+        return true;
+    }
+
+    ///
+    /// Append the ALTO XML for the layout of the image
+    ///
+    bool TessAltoRenderer::AddImageHandler(TessBaseAPI* api) {
+        const std::unique_ptr<const char[]> hocr(api->GetAltoText(imagenum()));
+        if (hocr == nullptr) return false;
+
+        AppendString(hocr.get());
+
+        return true;
+    }
+
+    ///
+    /// Append the ALTO XML for the end of the document
+    ///
+    bool TessAltoRenderer::EndDocumentHandler() {
+        AppendString("\t</Layout>\n</alto>\n");
+
+        return true;
+    }
+
+    TessAltoRenderer::TessAltoRenderer(const char *outputbase)
+        : TessResultRenderer(outputbase, "xml") {
+    }
+
+    ///
+    /// Make an XML-formatted string with ALTO markup from the internal
+    /// data structures.
+    ///
+    char *TessBaseAPI::GetAltoText(int page_number) {
+        return GetAltoText(nullptr, page_number);
+    }
+
+    ///
+    /// Make an XML-formatted string with ALTO markup from the internal
+    /// data structures.
+    ///
+    char *TessBaseAPI::GetAltoText(ETEXT_DESC *monitor, int page_number) {
+        if (tesseract_ == nullptr || (page_res_ == nullptr && Recognize(monitor) < 0))
+            return nullptr;
+
+        int lcnt = 0, bcnt = 0, wcnt = 0;
+        int page_id = page_number;
+
+        STRING alto_str("");
+
+        if (input_file_ == nullptr)
+            SetInputName(nullptr);
+
+        #ifdef _WIN32
+                // convert input name from ANSI encoding to utf-8
+              int str16_len =
+                  MultiByteToWideChar(CP_ACP, 0, input_file_->string(), -1, nullptr, 0);
+              wchar_t *uni16_str = new WCHAR[str16_len];
+              str16_len = MultiByteToWideChar(CP_ACP, 0, input_file_->string(), -1,
+                                              uni16_str, str16_len);
+              int utf8_len = WideCharToMultiByte(CP_UTF8, 0, uni16_str, str16_len, nullptr, 0,
+                                                 nullptr, nullptr);
+              char *utf8_str = new char[utf8_len];
+              WideCharToMultiByte(CP_UTF8, 0, uni16_str, str16_len, utf8_str,
+                                  utf8_len, nullptr, nullptr);
+              *input_file_ = utf8_str;
+              delete[] uni16_str;
+              delete[] utf8_str;
+        #endif
+
+            alto_str += "\t\t<Page WIDTH=\"";
+            alto_str.add_str_int("", rect_width_);
+            alto_str += "\" HEIGHT=\"";
+            alto_str.add_str_int("", rect_height_);
+            alto_str += "\" PHYSICAL_IMG_NR=\"";
+            alto_str.add_str_int("", rect_height_);
```

Comment:
@jakesebright , `rect_height` does not look like a `PHYSICAL_IMG_NR`. Is this a copy+paste error? What was intended here?

Reply:
I have a fix in pull request #2122.